src/actions.sol

```diff
@@ -58,8 +66,31 @@ contract Actions {
     event ApproveERC20(address indexed erc20, address indexed usr, uint256 amount);
     event TransferERC20(address indexed erc20, address indexed dst, uint256 amount);
 
+    address public immutable shelf;
+    address public immutable pile;
+    address public immutable self;
+
+    // address to deposit withdraws
+    address public immutable withdrawAddress;
+
+    // modifier only delegated call
+    modifier onlyDelegateCall() {
+        require(address(this) != self, "only-delegate-call");
```

Comment:
Why doesn't this fail? delegate calls use the full context of their caller?

Reply:
`self` is a immutable variable which is set in the constructor. It is the real address of the `tinlake-action` contract. 
It is immutable so not part of the storage (no impact in case of a delegat call).

In the delegate call `address(this)` will be the proxy address.

If tinlake-actions is called directly both address will be the same and it will revert